exclude_col parameter in CSVDataset.

stjohnso98 opened the pull request:
Resolves #39069.
Adds an optional parameter exclude_col in tf.data.experimental.CsvDataset.
Signed-off-by: Steenu Johnson <steenu.johnson@students.iiserpune.ac.in>

stjohnso98 committed:
Exclude col property in CSVDataset.

Signed-off-by: Steenu Johnson <steenu.johnson@students.iiserpune.ac.in>

jsimsa commented:
@rachellim could you please review? thanks

rachellim commented:
`exclude_cols` and `select_cols` are complements, IIUC. Is there a reason for adding an additional attribute to CsvDataset instead of just adding this in the python layer, something like the following:

```
exclude_cols_set = set(exclude_cols)
total_cols = len(record_defaults) + len(exclude_cols)
select_cols = [i for i in range(total_cols) if i not in exclude_cols_set]
```
?

rachellim reviewed (commented):
exclude_cols and select_cols are complements, IIUC. Is there a reason for adding an additional attribute to CsvDataset instead of just adding this in the python layer, something like the following:

```
exclude_cols_set = set(exclude_cols)
total_cols = len(record_defaults) + len(exclude_cols)
select_cols = [i for i in range(total_cols) if i not in exclude_cols_set]
```
?

stjohnso98 commented:
@rachellim Yes. If total_cols is calculated from user specified record_defaults and exclude_cols, some possible errors are not handled.
For example, if total_cols as calculated above is less than the actual number of columns, and the len(select_cols) equals len(record_default), all columns>total_cols will be excluded. No errors will be raised but the purpose of exclude_cols is not met in this case.

rachellim reviewed (commented):
The expected output shape of the dataset is computed based on record_defaults (see [1]). If, in the example you mentioned, the user intends to select more columns than is given in `record_defaults`, then the dataset's output spec will be wrong, which will result in errors when you try to iterate over the dataset. I'd encourage you to add a unit test for the case you mentioned to see what happens.


[1] https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/readers.py#L713-L714

stjohnso98 commented:
@rachellim Thanks for your suggestion. I am working on it. I have a couple of questions here. 
1.Suppose the CSV have 10 columns, all having integer values. User specifies record defaults = [tf.int32]*3 and exclude_cols = [2,3]. What is the expected behaviour of CSVDataset in this case? Should it throw an error or only select columns [1,4,5]?
 2. If it selects columns [1,4,5], then select_cols and exclude_cols will be mutually exclusive but would not be complements. Is this fine? Or if it does throw an error, could you please point me to the implementation where this is being handled? 
Thanks for your time again.

rachellim reviewed (commented):
Thanks for your contribution :) That's a good question. I think it should raise an error, because otherwise we'd be excluding more columns than the user specifies. 

That said, I wonder about the usefulness of an `exclude_cols` parameter, since the user is required to know the number and types of included columns anyway (`record_defaults`) and can easily compute `select_cols` knowing what columns they want to exclude. I think we should check on the original issue what the user's expectations are (are they hoping to not have to specify the included columns at all?).

Alternatively, we can have `exclude_cols` be a parameter on `make_csv_dataset` instead of `CsvDataset`. In `make_csv_dataset`,`record_defaults` is optional -- if it's not provided, we infer it from the first lines of the file. I think it'll be more useful for that scenario. What do you think?

stjohnso98 commented:
@rachellim Yes. I agree. exclude_cols would be more useful if the user do not have to specify record_defaults. Can you please check the original issue to confirm user expectations and guide me through the next steps?

gbaned commented:
@rachellim Can you please assist on above comments from @stjohnso98. Thanks!

rachellim commented:
Yup - @stjohnso98, I've commented on the original issue at #39069 -- feel free to add to that discussion as well if you have further thoughts. Thanks!

rachellim commented:
Looks like we've gotten a response on the original issue. I think adding `exclude_cols` as a python param on CsvDataset makes sense, and we should raise an error if the number of columns in the CSV file doesn't match the expected based on `record_defaults`. I believe CsvDataset already does this error checking, but would be good to double check (e.g. adding a unit test for this case if it doesn't already exist). 

@stjohnso98 let me know if you have further questions!

stjohnso98 commented:
@rachellim Sorry. I am a bit confused. In CsvDataset, when select_cols is specified, len(record_defaults) should equal len(select_cols). It raises an error only when this criteria is not met. In the above case I mentioned, since len(record_defaults) equals len(select_cols), it doesn't raise an error.

rachellim reviewed (commented):
You're right. I understand what you meant originally in https://github.com/tensorflow/tensorflow/pull/39799#issuecomment-634440978 now. 

It seems like we have two options:

1) Implement exclude_cols as a python "set complement", but in the case where the total_csv_columns_in_file > len(record_defaults) + len(exclude_cols), we'll exclude more columns than the user intends.

2) Implement exclude_cols as a CsvDatase attr. In the case where total_csv_columns_in_file > len(record_defaults) + len(exclude_cols), raise an error. 

Doing (2) makes sense to me, and I see now that that's what your original PR intended. I will go ahead and review it again.

Please also add (A) documentation in the docstring that this will raise an error if the total number of columns in the file exceeds len(record_defaults) + len(exclude_cols), and (B) a test that checks that this error is raised.

Thanks!

rachellim reviewed (changes requested):
Thanks for your PR! As mentioned before, please also add a unit test for the case where len(record_defaults) + len(exclude_cols) != number_of_columns_in_the_file. Also, add documentation for the error case in the python docstring, something along the lines of:

```
Raises:
  InvalidArgumentError: If `select_cols` is not None and 
    `len(select_cols) + len(record_defaults)` does not match the number of columns 
     in the file(s).
```

rachellim commented on tensorflow/python/data/experimental/ops/readers.py:
```diff
@@ -722,6 +732,7 @@ def __init__(self,
         use_quote_delim=self._use_quote_delim,
         na_value=self._na_value,
         select_cols=self._select_cols,
+        exclude_cols=self._exclude_cols,
```
(Related to above comment of creating a new version of this op)

For forward compatibility reasons, here we'll need to do something along the lines of:

```
if compat.forward_compatible(2020, 06, 25):
  variant_tensor = gen_experimental_dataset_ops.csv_dataset_v2(...)
else:
  variant_tensor = gen_experimental_dataset_ops.csv_dataset(...)
```

We maintain a 3 week forward compat window on tensorflow. https://www.tensorflow.org/api_docs/python/tf/compat/forward_compatible

stjohnso98 committed:
This commit creates a new op CSVDatasetV2 with an additional parameter exclude_cols for backwards compatibility.
Other changes include:
  Making the new op forward compatible.
  Adding error info to the docstring.
  Adding couple of tests to verify errors are being raised.

Signed-off-by: Steenu Johnson <steenu.johnson@students.iiserpune.ac.in>

stjohnso98 commented:
@rachellim Thanks for your comments. I have made the requested changes. Can you please review it again? Please let me know if there is anything wrong or in case I am missing something regarding the backward and forward compatibility changes. 
Also, I am not clear on the 3 weeks forward compatibility check. Could you please tell me what is expected of CSVDataset and exclude_cols when the forward compatibility window has not expired?

rachellim commented on tensorflow/core/kernels/data/experimental/csv_dataset_op.cc:
```diff
@@ -62,6 +64,14 @@ class CSVDatasetOp : public DatasetOpKernel {
     OP_REQUIRES(ctx, select_cols_tensor->dims() == 1,
                 errors::InvalidArgument("`select_cols` must be a vector."));
 
+    const Tensor* exclude_cols_tensor = new const Tensor();
+    if (op_version_ > 1) {
+      OP_REQUIRES_OK(ctx, ctx->input("exclude_cols", &exclude_cols_tensor));
+    }
```
```suggestion
std::vector<int64> exclude_cols;
if (op_version_ > 1) {
  const Tensor* exclude_cols_tensor;
  OP_REQUIRES_OK(ctx, ctx->input("exclude_cols", &exclude_cols_tensor));
  exclude_cols.reserve(exclude_cols_tensor->NumElements());
  for (int i = 0; i < exclude_cols_tensor->NumElements(); ++i) {
    exclude_cols.push_back(exclude_cols_tensor->flat<int64>()(i));
  }
}
```

Otherwise, in the v2 case, you're allocating a tensor (`new const Tensor()`) that you don't free.

rachellim reviewed (commented):
Besides the one comment, this looks good to me.

rachellim commented:
The forward compatibility window largely applies to the nightly build (and the internal build within google). When the window hasn't expired, we go through the `else` codepath, i.e. use the old version of the op.

stjohnso98 reviewed (commented):
Thanks for your suggestion. I will commit this suggestion. I have a couple of questions more.

rachellim reviewed (commented):
Thanks!

stjohnso98 committed:
Update tensorflow/core/kernels/data/experimental/csv_dataset_op.cc

Co-authored-by: Rachel Lim <lim.rachelys@gmail.com>

stjohnso98 commented on tensorflow/python/data/experimental/ops/readers.py:
```diff
@@ -710,19 +724,39 @@ def __init__(self,
-        na_value=self._na_value,
-        select_cols=self._select_cols,
-        compression_type=self._compression_type)
+    if compat.forward_compatible(2020, 6, 25):
+      variant_tensor = gen_experimental_dataset_ops.csv_dataset_v2(
+          filenames=self._filenames,
+          record_defaults=self._record_defaults,
+          buffer_size=self._buffer_size,
+          header=self._header,
+          output_shapes=self._flat_shapes,
+          field_delim=self._field_delim,
+          use_quote_delim=self._use_quote_delim,
+          na_value=self._na_value,
+          select_cols=self._select_cols,
+          exclude_cols=self._exclude_cols,
+          compression_type=self._compression_type)
+    else:
+      variant_tensor = gen_experimental_dataset_ops.csv_dataset(
+          filenames=self._filenames,
+          record_defaults=self._record_defaults,
+          buffer_size=self._buffer_size,
+          header=self._header,
+          output_shapes=self._flat_shapes,
+          field_delim=self._field_delim,
+          use_quote_delim=self._use_quote_delim,
+          na_value=self._na_value,
+          select_cols=self._select_cols,
+          compression_type=self._compression_type)
```
Yes. Shall I commit this change of yours as well? Also I committed your previous suggestion and googlebot has set cla:no. Can you please look into this as well? Thanks

rachellim replied:
Feel free to commit this suggestion, or directly modify the code yourself to do the same thing. I'll look into the CLA issue - for some reason github is associating my suggestion with the wrong email address...

rachellim commented on tensorflow/core/kernels/data/experimental/csv_dataset_op.cc:
```diff
@@ -62,6 +64,19 @@ class CSVDatasetOp : public DatasetOpKernel {
     OP_REQUIRES(ctx, select_cols_tensor->dims() == 1,
                 errors::InvalidArgument("`select_cols` must be a vector."));
 
+std::vector<int64> exclude_cols;
```
You'll need to format this with the right indentation.

rachellim commented on tensorflow/core/kernels/data/experimental/csv_dataset_op.cc:
```diff
@@ -28,7 +28,9 @@ namespace {
-  explicit CSVDatasetOp(OpKernelConstruction* ctx) : DatasetOpKernel(ctx) {
+  explicit CSVDatasetOp(OpKernelConstruction* ctx)
+      : DatasetOpKernel(ctx),
+        op_version_(ctx->def().op() == "CSVDataset" ? 1 : 2) {
```
I don't see this change, maybe you forgot to commit?

rachellim commented on tensorflow/core/kernels/data/experimental/csv_dataset_op.cc:
```diff
@@ -126,11 +141,29 @@ class CSVDatasetOp : public DatasetOpKernel {
-                          output_types_, output_shapes_,
-                          std::move(record_defaults), std::move(select_cols),
-                          use_quote_delim, delim[0], std::move(na_value));
+    std::vector<int64> exclude_cols;
+    exclude_cols.reserve(exclude_cols_tensor->NumElements());
+    for (int i = 0; i < exclude_cols_tensor->NumElements(); ++i) {
+      exclude_cols.push_back(exclude_cols_tensor->flat<int64>()(i));
+    }
```
You'll need to remove this too (this clashes with the above suggestion you committed)

rachellim commented on tensorflow/core/kernels/data/experimental/csv_dataset_op.cc:
```diff
@@ -238,8 +243,8 @@ class CSVDatasetOp : public DatasetOpKernel {
            std::make_pair(6, na_value), std::make_pair(7, select_cols),
-           std::make_pair(8, exclude_cols)},     // Single tensor inputs
-          {std::make_pair(9, record_defaults)},  // Tensor list inputs
+           std::make_pair(9, exclude_cols)},     // Single tensor inputs
```
I don't see this change, maybe you forgot to commit?

rachellim commented on tensorflow/python/data/experimental/ops/readers.py:
```diff
@@ -722,18 +732,31 @@ def __init__(self,
-        select_cols=self._select_cols,
-        exclude_cols=self._exclude_cols,
-        compression_type=self._compression_type)
+    if compat.forward_compatible(2020, 6, 25):
```
bump

rachellim reviewed (changes requested):
Thanks!

rachellim commented:
*Manually updated cla tag*

stjohnso98 committed:
Forward compatiblity and backwards compatibility changes.

Changing forward compatibility date to be 3 weeks from the current commit.
Using V2 version when user specifies exclude_cols.
Making op_version in kernel more specific to CSVDatasetV2.
Serialization done only for version 2 op.
Removing clashes from the previous commit and fixing indentation errors.

Signed-off-by: Steenu Johnson <steenu.johnson@students.iiserpune.ac.in>

stjohnso98 commented:
@rachellim Thanks for your review. I have made the other changes as well. Can you please review it again?
Also, the cla issue happened again. 
Thanks.

rachellim reviewed (dismissed):
Thank you!

rachellim commented:
Overrided CLA status because the github commit was associated with the wrong email (my personal email instead of my google.com email)

rachellim commented:
@stjohnso98, looks like some checks are failing. You'll have to add an api_def_CSVDatasetV2.pbtxt file for the failing `//tensorflow/core/api_def:api_test` test. You can base it on https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/base_api/api_def_CSVDataset.pbtxt

You'll also have to update golden apis for the api_compatibility_test to pass. You can do so as follows:
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

Thanks!

stjohnso98 committed:
Fixing build errors.

api_test
api_compatibility_test

Signed-off-by: Steenu Johnson <steenu.johnson@students.iiserpune.ac.in>

stjohnso98 commented:
@rachellim Thanks for your comment. I have made the suggested changes. Can you please review it again?
Also cla status has changed. Apologies to bring it up again.

rachellim commented:
No worries! Thanks for the fix :)

stjohnso98 commented:
All checks have passed. Why is the merging delayed?

gbaned commented:
> All checks have passed. Why is the merging delayed?

@stjohnso98  It is processing internally. Thanks!
